goon/code/datums/browserOutput.dm

```diff
@@ -0,0 +1,306 @@
+	var/cookieSent   = 0 // Has the client sent a cookie for analysis
+	var/list/connectionHistory = list() //Contains the connection history passed from chat cookie
+	var/broken       = FALSE
+
+/datum/chatOutput/New(client/C)
+	. = ..()
+
+	owner = C
+	// world.log << "chatOutput: New()"
+
+/datum/chatOutput/proc/start()
+	//Check for existing chat
+	if(!owner)
+		return 0
+
+	if(!winexists(owner, "browseroutput")) // Oh goddamnit.
+		alert(owner.mob, "Updated chat window does not exist. If you are using a custom skin file please allow the game to update.")
+		broken = TRUE
+		return 0
+
+	if(winget(owner, "browseroutput", "is-disabled") == "false") //Already setup
+		doneLoading()
+
+	else //Not setup
+		load()
+
+	return 1
+
+/datum/chatOutput/proc/load()
+	set waitfor = FALSE
+	if(!owner)
+		return
+
+	//world.log << "chatOutput: load()"
+
+	for(var/attempts = 1 to 5)
+		for(var/asset in global.chatResources) // No asset cache, just get this fucking shit SENT.
+			owner << browse_rsc(file(asset))
+
+		//world.log << "Sending main chat window to client [owner.ckey]"
+		owner << browse(file("goon/browserassets/html/browserOutput.html"), "window=browseroutput")
+		sleep(20 SECONDS)
+		if(!owner || loaded)
+			break
+
+	//world.log << "chatOutput: load() completed"
+
+/datum/chatOutput/Topic(var/href, var/list/href_list)
+	if(usr.client != owner)
+		return 1
+	// Build arguments.
+	// Arguments are in the form "param[paramname]=thing"
+	var/list/params = list()
+	for(var/key in href_list)
+		if(length(key) > 7 && findtext(key, "param")) // 7 is the amount of characters in the basic param key template.
+			var/param_name = copytext(key, 7, -1)
+			var/item       = href_list[key]
+
+			params[param_name] = item
+
+	var/data // Data to be sent back to the chat.
+	switch(href_list["proc"])
+		if("doneLoading")
+			data = doneLoading(arglist(params))
+
+		if("debug")
+			data = debug(arglist(params))
+
+		if("ping")
+			data = ping(arglist(params))
+
+		if("analyzeClientData")
+			data = analyzeClientData(arglist(params))
+
+		if("encoding")
+			var/encoding = href_list["encoding"]
+			var/static/regex/RE = regex("windows-(874|125\[0-8])")
+			if (RE.Find(encoding))
+				owner.encoding = RE.group[1]
+
+			else if (encoding == "gb2312")
+				owner.encoding = "2312"
+
+			// This seems to be the result on Japanese locales, but the client still seems to accept 1252.
+			else if (encoding == "_autodetect")
+				owner.encoding = "1252"
+
+			else
+				stack_trace("Unknown encoding received from client: \"[sanitize(encoding)]\". Please report this as a bug.")
+
+	if(data)
+		ehjax_send(data = data)
+
+//Called on chat output done-loading by JS.
+/datum/chatOutput/proc/doneLoading()
+	if(loaded)
+		return
+
+	loaded = TRUE
+	winset(owner, "browseroutput", "is-disabled=false")
+	for(var/message in messageQueue)
+		to_chat(owner, message)
+
+	messageQueue = null
+	src.sendClientData()
```

Comment:
Could you use /tg/'s version? It's pretty similar to /vg/'s, but already cleaned up.

Reply:
Their version of what?

Reply:
Of goonchat. The same file in their repo would be this: https://github.com/tgstation/tgstation/blob/master/code/modules/goonchat/browserOutput.dm

Reply:
At some point in the future sure but this is something i know works with this codebase without any additional work.

Reply:
Fair enough.